bokehjs/src/coffee/embed.ts

```diff
@@ -0,0 +1,336 @@
+import * as base from "./base"
+import {logger, set_log_level} from "./core/logging"
+import {Document, DocJson, DocumentChangedEvent, RootAddedEvent, RootRemovedEvent, TitleChangedEvent} from "./document"
+import {div, link, style, replaceWith} from "./core/dom"
+import {defer} from "./core/util/callback"
+import {unescape} from "./core/util/string"
+import {size, values} from "./core/util/object"
+import {isString} from "./core/util/types"
+import {Receiver} from "./protocol/receiver"
+import {Message} from "./protocol/message"
+import {ClientSession} from "./client/session"
+import {pull_session} from "./client/connection"
+import {HasProps} from "./core/has_props"
+import {DOMView} from "./core/dom_view"
+
+export type DocsJson = {[key: string]: DocJson}
+
+export interface RenderItem {
+  elementid: string
+  docid?: string
+  modelid?: string
+  sessionid?: string
+  use_for_title?: boolean
+  notebook_comms_target?: any
+}
+
+declare interface Comm {
+  target_name: string
+  on_msg: (...args: any[]) => void
+}
+
+declare interface Jupyter {
+  notebook: {
+    kernel: undefined | {
+      comm_manager: {
+        comms: {[key: string]: Promise<Comm>},
+        register_target: (target: string, fn: (comm: Comm) => void) => void,
+      }
+    }
+  }
+}
+
+declare var Jupyter: Jupyter | undefined
+
+// Matches Bokeh CSS class selector. Setting all Bokeh parent element class names
+// with this var prevents user configurations where css styling is unset.
+export const BOKEH_ROOT = "bk-root"
+
+function _handle_notebook_comms(this: Document, receiver: Receiver, msg?: Message | null): void {
+  if (msg != null) {
+    if (msg.buffers.length > 0)
+      receiver.consume((msg.buffers[0] as any).buffer) // XXX: buffers[i] is a tuple
```

Comment:
@bryevdv, can you look at this function and this line in particular, because either this code or my understanding of involved types is wrong.

Reply:
@mattpap the message is a jupyter comms message which wraps another layer around things:
<img width="1332" alt="screen shot 2017-12-13 at 10 57 25" src="https://user-images.githubusercontent.com/1078448/33951425-89bf9c48-dff4-11e7-8d4d-9d08c43e8e35.png">

This code is to discern between json vs binary messages given how jupyter comms messages are structured.

Reply:
when we send a binary array it is a single `DataView` in the `.buffers` list:
<img width="1325" alt="screen shot 2017-12-13 at 11 01 57" src="https://user-images.githubusercontent.com/1078448/33951636-1eee9990-dff5-11e7-920f-c24e0bd6aa57.png">